core/src/main/scala/org/genivi/sota/core/daemon/DeltaListener.scala

```diff
@@ -0,0 +1,62 @@
+package org.genivi.sota.core.daemon
+
+import akka.Done
+import org.genivi.sota.common.DeviceRegistry
+import org.genivi.sota.core.UpdateService
+import org.genivi.sota.core.campaigns.CampaignLauncher
+import org.genivi.sota.core.db.{Campaigns, Packages}
+import org.genivi.sota.messaging.MessageBusPublisher
+import org.genivi.sota.messaging.Messages.{DeltaGenerationFailed, GeneratedDelta}
+import slick.driver.MySQLDriver.api._
+import org.genivi.sota.core.data.Campaign
+import org.slf4j.LoggerFactory
+import org.genivi.sota.messaging.Commit.Commit
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class DeltaListener(deviceRegistry: DeviceRegistry, updateService: UpdateService,
+                    messageBus: MessageBusPublisher)(implicit db: Database)  {
+
+  private val log = LoggerFactory.getLogger(this.getClass)
+
+  private def validateMessage(campaign: Campaign, from: Commit, to: Commit)
+                             (implicit ec: ExecutionContext): DBIO[Unit] = {
+    //this method only returns a DBIO so it can be used inside a for comprehension involving slick calls
+    if(campaign.meta.deltaFrom.isEmpty) {
+      DBIO.failed(new IllegalArgumentException("Received GeneratedDelta message for campaign without static delta"))
+    } else if(campaign.meta.deltaFrom.get.version.get.equalsIgnoreCase(from.get)) {
+      DBIO.failed(
+        new IllegalArgumentException("Received GeneratedDelta message for campaign with differing from version"))
+    } else if(campaign.meta.packageUuid.isEmpty) {
+      DBIO.failed(new IllegalArgumentException("Received GeneratedDelta message for campaign without a target version"))
+    } else {
+        Packages.byUuid(campaign.meta.packageUuid.get.toJava).map { pkg =>
+          if (pkg.id.version.get.equalsIgnoreCase(to.get)) {
+            DBIO.failed(new IllegalArgumentException(s"Version in GeneratedDelta message ($to) doesn't match version " +
+              s"in campaign (${pkg.id.version})"))
+          } else {
+            DBIO.failed(new IllegalArgumentException(s"Failed to read package for campaign"))
+          }
```

Comment:
Ahh, this `Unit` business in Scala bites us far more than it deserves. These `DBIO.failed` will be ignored (a silent `; Unit`). So you want to use a `flatMap` rather than a `map`, but maybe also change the `Unit` to `Done` to prevent this kind of silly bugs (that Scala shouldn't accept).